### DISCUSSION TITLE
CSS: create a responsive grid layout using only CSS

### DISCUSSION THREAD

**Discussion by vladimir-de-oppenheimer** [community/community#138562] [Programming Help]
**How can you create a responsive grid layout using only CSS, ensuring that the grid items maintain a 16:9 aspect ratio, and the grid itself adjusts to different screen sizes without using JavaScript or fixed heights?**

**Comment by queenofcorgis** [comment]
Hey there!  👋 

Thanks for posting in the GitHub Community, @vladimir-de-oppenheimer ! We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category. The Accessibility category is a place for our community to discuss and provide feedback on the digital accessibility of GitHub products. Digital accessibility means that GitHub tools, and technologies, are designed and developed so that people with disabilities can use them.

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.

**Reply by jgschmitz** [reply]
I haven't looked at this stuff in a while but I think the key to maintaining the 16:9 aspect ratio is using padding based on a percentage of the width of the element. In a 16:9 ratio, the height is 56.25% of the width (i.e., 9/16 = 0.5625 or 56.25%). 
Here is a code example
``` html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Responsive 16:9 Grid</title>
  <style>
    /* Create a responsive grid container */
    .grid-container {
      display: grid;
      grid-template-columns: repeat(auto-fill, minmax(250px, 1fr)); /* Responsive column layout */
      gap: 10px; /* Spacing between grid items */
      padding: 10px;
    }

    /* Grid item to maintain 16:9 aspect ratio */
    .grid-item {
      position: relative; /* Needed for the padding trick */
      background-color: #ddd;
    }

    /* Create a 16:9 aspect ratio using ::before pseudo-element */
    .grid-item::before {
      content: "";
      display: block;
      padding-top: 56.25%; /* 16:9 Aspect ratio (9/16 * 100 = 56.25) */
    }

    /* Content inside the grid item */
    .grid-item-content {
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      display: flex;
      justify-content: center;
      align-items: center;
      background-color: rgba(0, 0, 0, 0.5);
      color: white;
      font-size: 24px;
    }
  </style>
</head>
<body>

  <div class="grid-container">
    <div class="grid-item">
      <div class="grid-item-content">1</div>
    </div>
    <div class="grid-item">
      <div class="grid-item-content">2</div>
    </div>
    <div class="grid-item">
      <div class="grid-item-content">3</div>
    </div>
    <div class="grid-item">
      <div class="grid-item-content">4</div>
    </div>
    <div class="grid-item">
      <div class="grid-item-content">5</div>
    </div>
  </div>

</body>
</html>
```

Hope this helps! cheers 🍻

**Comment by LipeLacross** [comment]
To create a responsive grid layout that maintains a 16:9 aspect ratio for grid items using only CSS, you can utilize the "padding-bottom" technique along with CSS Grid or Flexbox. Here’s a detailed example using CSS Grid:

```
html
Copy code
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Responsive Grid Layout</title>
    <style>
        body {
            margin: 0;
            padding: 0;
            display: flex;
            justify-content: center;
            align-items: center;
            min-height: 100vh;
            background-color: #f0f0f0;
        }

        .grid-container {
            display: grid;
            grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
            gap: 10px;
            width: 90vw;
            max-width: 1200px;
        }

        .grid-item {
            position: relative;
            padding-bottom: 56.25%; /* 16:9 Aspect Ratio */
            background-color: #ccc;
            border: 1px solid #ddd;
            overflow: hidden;
        }

        .grid-item img {
            position: absolute;
            top: 0;
            left: 0;
            width: 100%;
            height: 100%;
            object-fit: cover;
        }
    </style>
</head>
<body>
    <div class="grid-container">
        <div class="grid-item">
            <img src="https://via.placeholder.com/600x337" alt="Placeholder Image">
        </div>
        <div class="grid-item">
            <img src="https://via.placeholder.com/600x337" alt="Placeholder Image">
        </div>
        <div class="grid-item">
            <img src="https://via.placeholder.com/600x337" alt="Placeholder Image">
        </div>
        <!-- Add more grid items as needed -->
    </div>
</body>
</html>
```
Detailed Explanation:
Grid Container Setup:

display: grid;: Defines the container as a grid.
grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));: Creates responsive columns. auto-fill allows the number of columns to adjust based on the available space, while minmax(200px, 1fr) ensures each column has a minimum size of 200px and takes up a fraction of the remaining space.
Maintaining 16:9 Aspect Ratio:

padding-bottom: 56.25%;: This technique uses the bottom padding to create a box with the desired aspect ratio. The percentage is calculated based on the formula height / width * 100%, where 16:9 translates to 56.25% (9 / 16 * 100%). This makes the bottom padding of .grid-item determine its height, maintaining the 16:9 ratio.
Image Positioning:

.grid-item: position: relative; allows the padding-bottom to set the height, and the image inside is positioned absolutely.
.grid-item img: position: absolute; ensures the image covers the entire space of the grid item. object-fit: cover; makes sure the image fills the container without distortion, cropping parts of the image if necessary to maintain the ratio.
Responsiveness:

The grid adjusts automatically to the screen size due to auto-fill and minmax, ensuring the layout is flexible and responsive. The container's maximum width (max-width: 1200px;) and automatic width adjustment (width: 90vw;) ensure the layout adapts well to different screen sizes.
This approach uses only CSS and does not require JavaScript, ensuring a responsive layout with a consistent aspect ratio for the grid items.

**Comment by sabbirsami** [comment]
# Responsive Grid Layout with 16:9 Aspect Ratio Using Only CSS

This example demonstrates how to create a responsive grid layout that ensures all grid items maintain a 16:9 aspect ratio. The layout adjusts to different screen sizes without using JavaScript or fixed heights.

## Key Concepts:
- The grid is defined using CSS Grid Layout.
- The `padding-bottom` technique is used to maintain a 16:9 aspect ratio.
- Media queries ensure the grid layout is responsive across different screen sizes.

## HTML and CSS Solution:

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Responsive Grid with 16:9 Aspect Ratio</title>
  <style>
    /* Base grid layout */
    .grid-container {
      display: grid;
      gap: 1rem;
      grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
      padding: 1rem;
    }

    /* Grid items with 16:9 aspect ratio */
    .grid-item {
      position: relative;
      background-color: #ccc;
      overflow: hidden;
    }

    .grid-item::before {
      content: "";
      display: block;
      padding-bottom: 56.25%; /* 16:9 aspect ratio (9 / 16 = 0.5625) */
    }

    .grid-item-content {
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      display: flex;
      justify-content: center;
      align-items: center;
      background-color: rgba(0, 0, 0, 0.5);
      color: white;
      font-size: 1.5rem;
    }

    /* Responsive design */
    @media (max-width: 600px) {
      .grid-container {
        grid-template-columns: 1fr;
      }
    }
  </style>
</head>
<body>

  <div class="grid-container">
    <div class="grid-item">
      <div class="grid-item-content">Item 1</div>
    </div>
    <div class="grid-item">
      <div class="grid-item-content">Item 2</div>
    </div>
    <div class="grid-item">
      <div class="grid-item-content">Item 3</div>
    </div>
    <div class="grid-item">
      <div class="grid-item-content">Item 4</div>
    </div>
  </div>

</body>
</html>
```

**Comment by ebndev** [comment]
Please see [this comment](https://github.com/orgs/community/discussions/139491#discussioncomment-10730360) for reference to why this discussion was closed and locked. Thanks!
